indexbs/indexbacked_bs.go

```diff
@@ -188,42 +179,13 @@ func (ro *IndexBackedBlockstore) execOp(ctx context.Context, c cid.Cid, op Block
 		return nil, fmt.Errorf("failed to load read-only blockstore for shard %s: %w", sk, err)
 	}
 
-	// Call the operation on the blockstore
-	res, err = execOpOnBlockstore(ctx, c, sk, bs, op)
-	if err != nil {
-		return nil, err
-	}
-
-	// Update the blockstore cache
+	// Add the blockstore to the cache
 	ro.blockstoreCache.Add(sk, &accessorWithBlockstore{sa, bs})
```

Comment:
@dirkmc If multiple threads fail to get the blockstore above, they will ALL call Acquire Shard which in turn will increment the reference count for the shard access as many times. However, those reference counts will then not get released by calling Close on them(since that will only happen once on eviction). Thus, the other blockstores will leak.

Reply:
This is why the entire (Get + Add) i.e. _GetOrCreateIfNotExistAndGet_ dance needs to be atomic.